### ISSUE TITLE
Rework "serverless rollback" support

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#3894]
# This is a Feature Proposal

## Description

A while back [we've introduce](https://github.com/serverless/serverless/pull/2495) rollback support for serverless deployments with the help of the `serverless deploy list` and `serverless rollback --timestamp` commands.

This supports enables a way to take a look at all the deployment artifacts which are stored in the deployment bucket (with the help of `serverless deploy list`) and then rollback to a specific deployment based on the timestamp of the deployment (with the help of the `serverless rollback --timestamp` command).

While this is a good starting point it is worthwhile to re-think the whole behavior and functionality and add support for more feature rich deployment rollbacks.

This issue discusses potential improvements and current issues we should fix in an upcoming update.

**Current issues:**

- It's currently not visible which version of the service is currently deployed when running `serverless deploy list`
- Rollbacks can't be automated due to hard to parse CLI output (see [here](https://github.com/serverless/serverless/issues/2412#issuecomment-306981467) for more info).
- Rollback doesn't re-upload the files which will break the "Skip if files not changed" functionality (because it always compares to the most recent deployment)
- Picking a version to rollback to based on the timestamp is not a great DX

**Feature wishlist:**

- Add support to rollback to previous versions with the help of a `number` options (such as `serverless rollback -n 4` --> Rollback 4 versions in time)
- Parseable output? (discussed and implemented in https://github.com/serverless/serverless/issues/3089)
- ...

/cc @brianneisler @eahefnawy

**Comment by evan-king** [comment]
As long as rolling back by timestamp is supported, I would like to be able to supply an inexact input, such that Serverless will choose the newest version whose timestamp does not exceed the one specified in rollback command.

I would especially like to be able to handle list and rollback operations with API calls alternatively to CLI invocation.

**Comment by pmuens** [comment]
Nice! Thanks for the detailed feedback @evan-king 👍 

> As long as rolling back by timestamp is supported, I would like to be able to supply an inexact input, such that Serverless will choose the newest version whose timestamp does not exceed the one specified in rollback command.

That sounds like a good addition. Would you use this over the `--number` option (where you could go back e.g. one deployment when providing `-n 1`)?

> I would especially like to be able to handle list and rollback operations with API calls alternatively to CLI invocation.

Do you mean calling the Framework programmatically via an exposed Framework API? So basically a replacement to run `serverless` commands in a sub process when e.g. running it in CI?

**Comment by pmuens** [comment]
Friendly pinging @serverless/vip for other, enterprise related features.

**Comment by evan-king** [comment]
In principle, I would prefer either an identifier or a timestamp over a number of rollouts, to minimize the potential for concurrency issues/leave openings for CD solutions that allow/cope with some forms of concurrency.

My reasoning:
> This consideration would almost assuredly indicate over-engineering for the scale of development I personally face supporting - hence the "in principle" qualifier.  But I think it's fair to think this a realistic concern if deployment automation includes long-tail monitoring, gradual switchover, etc.  Such scenarios might get a version into operation in minutes but take hours or even days to be totally "finished" with that deployment.  At such point, I think it definitely prudent to ensure smart handling of releases that are a blip between patches, superceded before they're even finished deploying - and that's an easier proposition when avoiding shared and/or mutable state wherever possible.  Overlapping deployments, for example, can more easily inherit the starting timestamp of the earliest in-progress (possibly up to a certain stage) deployment as its rollback target than figure out how many versions to roll back at the point of failure.  What if 2 deployments fail in parallel?

At the least, such considerations might allow simpler pipelines to avoid requiring some form of locking mechanism explicitly preventing concurrency.

And yes, I mean programmatic invocation of a documented Framework API.

**Comment by goncaloneves** [comment]
`deploy list` and `rollback` probably are the most important Serverless commands after `deploy`.

My idea is, adding a Serverless meta data information file to S3 deployments that could provide more info in the `deploy list` command such as:

- Serverless version
- Provider
- Region
- Stage
- Function versions
- Custom key/value (set in `serverless.yml`)
- Copy of the entire `serverless.yml` (?)

At `deploy` time this file could get custom data from `serverless.yml` if the developer wants to associate specific tags and key/value to a specific deployment.

**Comment by pmuens** [comment]
Thanks for commenting and giving feedback @goncaloneves 👍

I like the metadata idea 👍 Could be really helpful when implementing this feature!

**Aside:** We already use metadata to e.g. store the hash of the .zip file which is then used to compare the local version with he remotely deployed artifact and only re-deploy when they are different!

**Comment by jscattergood** [comment]
I currently implemented the scripted approach to perform an automatic rollback using Jenkins, but now find that the rollback doesn’t remove any files in the S3 bucket and I couldn’t find an API to delete a deployment.  Without a reference to the timestamp that’s actually deployed,  I can’t reliably determine which timestamp to rollback to.

Any thoughts?

**Comment by pmuens** [comment]
Closing since this issue is quite old and stale and we're currently not planning to work on this anytime soon.
